src/GitHub.App/Services/PullRequestService.cs

```diff
@@ -242,93 +271,92 @@ public IObservable<Unit> SwitchToBranch(ILocalRepositoryModel repository, IPullR
-            ILocalRepositoryModel repository,
-            IModelService modelService,
-            string commitSha,
-            string fileName,
-            string fileSha)
-        {
-            return Observable.Defer(async () =>
-            {
-                var repo = gitService.GetRepository(repository.LocalPath);
-                var remote = await gitClient.GetHttpRemote(repo, "origin");
-                await gitClient.Fetch(repo, remote.Name);
-                var result = await GetFileFromRepositoryOrApi(repository, repo, modelService, commitSha, fileName, fileSha);
-
-                if (result == null)
-                {
-                    throw new FileNotFoundException($"Could not retrieve {fileName}@{commitSha}");
-                }
-
-                return Observable.Return(result);
+                var configKey = string.Format(
+                    CultureInfo.InvariantCulture,
+                    "branch.{0}.{1}",
+                    repo.Head.FriendlyName,
+                    SettingGHfVSPullRequest);
+                return gitClient.GetConfig<int>(repo, configKey).ToObservable();
             });
         }
 
         public IObservable<Tuple<string, string>> ExtractDiffFiles(
             ILocalRepositoryModel repository,
-            IModelService modelService,
             IPullRequestModel pullRequest,
             string fileName,
-            string fileSha,
             bool isPullRequestBranchCheckedOut)
         {
             return Observable.Defer(async () =>
             {
                 var repo = gitService.GetRepository(repository.LocalPath);
                 var remote = await gitClient.GetHttpRemote(repo, "origin");
-                await gitClient.Fetch(repo, remote.Name);
-
-                // The left file is the target of the PR so this should already be fetched.
-                var left = await gitClient.ExtractFile(repo, pullRequest.Base.Sha, fileName);
-
-                // The right file - if it comes from a fork - may not be fetched so fall back to
-                // getting the file contents from the model service.
-                var right = isPullRequestBranchCheckedOut ?
-                    Path.Combine(repository.LocalPath, fileName) :
-                    await GetFileFromRepositoryOrApi(repository, repo, modelService, pullRequest.Head.Sha, fileName, fileSha);
 
-                if (left == null)
+                var baseSha = pullRequest.Base.Sha;
+                var headSha = pullRequest.Head.Sha;
+                var baseRef = pullRequest.Base.Ref;
+                string mergeBase = await gitClient.GetPullRequestMergeBase(repo, remote.Name, baseSha, headSha, baseRef, pullRequest.Number);
+                if (mergeBase == null)
                 {
-                    throw new FileNotFoundException($"Could not retrieve {fileName}@{pullRequest.Base.Sha}");
+                    throw new FileNotFoundException($"Couldn't find merge base between {baseSha} and {headSha}.");
```

Comment:
I was reluctant to create a new exception type for something that really shouldn't happen. I've changed it to `InvalidOperationException` which we're using elsewhere in this class. Still not ideal, but at least it isn't an IO exception.